kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/AccountManagerImpl.java

```diff
@@ -0,0 +1,123 @@
+	private final EventLogger eventLogger;
+	private AuthDataSource authRepository;
+	private KinAccount kinAccount;
+	private final ObservableData<Integer> accountState;
+
+	private ListenerRegistration accountCreationRegistration;
+
+	AccountManagerImpl(@NonNull final AccountManager.Local local, @NonNull final EventLogger eventLogger) {
+		this.local = local;
+		this.eventLogger = eventLogger;
+		this.accountState = ObservableData.create(NONE);
+	}
+
+	@Override
+	public @AccountState
+	int getAccountState() {
+		return local.getAccountState();
+	}
+
+	@Override
+	public void addAccountStateObserver(@NonNull Observer<Integer> observer) {
+		accountState.addObserver(observer);
+	}
+
+	@Override
+	public void removeAccountStateObserver(@NonNull Observer<Integer> observer) {
+		accountState.removeObserver(observer);
+	}
+
+
+	void start(@NonNull final AuthDataSource authRepository, @NonNull final KinAccount kinAccount) {
+		new Log().withTag(TAG).put("setAccountState", "start").log();
+		this.authRepository = authRepository;
+		this.kinAccount = kinAccount;
+		this.setAccountState(local.getAccountState());
+	}
+
+	private void setAccountState(@AccountState final int accountState) {
+		if (!isPreviousState(this.accountState.getValue(), accountState)) {
+			this.accountState.postValue(accountState);
+			this.local.setAccountState(accountState);
+			switch (accountState) {
+				case REQUIRE_CREATION:
+					eventLogger.send(StellarAccountCreationRequested.create());
+					new Log().withTag(TAG).put("setAccountState", "REQUIRE_CREATION").log();
+					// Trigger account creation from server side.
+					authRepository.getAuthToken(new KinCallbackAdapter<AuthToken>() {
+						@Override
+						public void onResponse(AuthToken response) {
+							setAccountState(PENDING_CREATION);
+						}
+					});
+					break;
+				case PENDING_CREATION:
+					new Log().withTag(TAG).put("setAccountState", "PENDING_CREATION").log();
+					// Start listen for account creation on the blockchain side.
+					accountCreationRegistration = kinAccount.blockchainEvents()
+						.addAccountCreationListener(new EventListener<Void>() {
+							@Override
+							public void onEvent(Void data) {
+								accountCreationRegistration.remove();
+								accountCreationRegistration = null;
+								setAccountState(REQUIRE_TRUSTLINE);
+							}
+						});
+					break;
+				case REQUIRE_TRUSTLINE:
+					new Log().withTag(TAG).put("setAccountState", "REQUIRE_TRUSTLINE").log();
+					// Create trustline transaction with KIN
+					new CreateTrustLineCall(kinAccount, new TrustlineCallback() {
+						@Override
+						public void onSuccess() {
+							eventLogger.send(StellarKinTrustlineSetupSucceeded.create());
+							setAccountState(CREATION_COMPLETED);
+						}
+
+						@Override
+						public void onFailure(OperationFailedException e) {
+							eventLogger.send(StellarKinTrustlineSetupFailed.create(e.getMessage()));
+							setAccountState(REQUIRE_TRUSTLINE);
```

Comment:
setAccountState(REQUIRE_TRUSTLINE); will have no affect as current state is already REQUIRE_TRUSTLINE

Reply:
isValidStateUpdate can contain a more detailed logic such as if current state = REQUIRE_TRUSTLINE and new sate is also REQUIRE_TRUSTLINE allow retry REQUIRE_TRUSTLINE

Reply:
need also to handle failure (so we can retry)